src/mlpack/methods/reinforcement_learning/replay/sumtree.hpp

```diff
@@ -0,0 +1,193 @@
+/**
+ * @file sumtree.hpp
+ * @author Xiaohong
+ *
+ * This file is an implementation of sumtree.
+ *
+ * reference:
+ * [1] https://github.com/openai/baselines/blob/master/baselines/common/segment_tree.py
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_SUMTREE_HPP
+#define MLPACK_METHODS_RL_SUMTREE_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * Implementation of SumTree.
+ *
+ * Build a Segment Tree like data structure.
+ * https://en.wikipedia.org/wiki/Segment_tree
+ *
+ * Used to maintain prefix-sum of an array.
+ *
+ */
+
+template<typename T>
+class SumTree
+{
+ public:
+
+  /**
+   * Default constructor
+   */
+  SumTree()
+  { /* Nothing to do here. */ }
+
+  /**
+   * Construct an instance of SumTree class
+   *
+   * @param capacity The size of data
+   */
+  SumTree(size_t capacity):
+      capacity(capacity)
+  {
+    element = std::vector<T> (2 * capacity);
+  }
+
+  /**
+   * Set the data array with idx
+   *
+   * @param idx  The array idx to be changed
+   * @param value  The data that array with idx to be
+   */
+  void set(size_t idx, T value)
+  {
+    idx += capacity;
+    element[idx] = value;
+    idx /= 2;
+    while (idx >= 1)
+    {
+      element[idx] = element[2 * idx] + element[2 * idx + 1];
+      idx /= 2;
+    }
+  }
+
+
+  /**
+   * Update the data with batch rather loop over the indices with set method
+   *
+   * @param indices The indices of data to be changed
+   * @param data  The data that array with indices to be
+   */
+  void batchUpdate(arma::ucolvec indices, arma::Col<T> data)
+  {
+    for (size_t i = 0; i < indices.n_rows; i ++)
+    {
+      element[indices[i] + capacity] = data[i];
+    }
+    // update the total tree with bottom-up technique
+    for (size_t i = capacity-1; i > 0; i --)
+    {
+      element[i] = element[2 * i] + element[2 * i + 1];
+    }
+  }
+
+  /**
+   * Get the data array with idx
+   *
+   * @param idx The array idx to get data
+   */
+  T get(size_t idx)
+  {
+    idx += capacity;
+    return element[idx];
+  }
+
+  /**
+   * Help function for the `sum` function
+   *
+   * @param _start The starting position of subsequence
+   * @param _end The end position of subsequence
+   * @param node Reference position
+   * @param node_start Starting position of reference segment
+   * @param node_end End position of reference segment
+   */
+  T sumHelper(size_t _start, size_t _end, size_t node, size_t node_start, size_t node_end)
+  {
+    if (_start == node_start && _end == node_end)
```

Comment:
Variable name should start with a lowercase letter. So please avoid underscore. Same for all other places.